File: Algorithms Advanced/others/GraphAlgorithms/C++/Prim.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

void prims(int nodes, vector<vector<pair<int, int>>> &adj) 
{
    vector<int> key(nodes, INT_MAX), parent(nodes, -1);
    vector<bool> inMST(nodes, false);
    key[0] = 0;

    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    pq.push({0, 0});

    while (!pq.empty()) 
    {
        int u = pq.top().second;
        pq.pop();

        inMST[u] = true;

        for (auto &[v, weight] : adj[u]) 
        {
            if (!inMST[v] && weight < key[v]) 
            {
                key[v] = weight;
                pq.push({key[v], v});
                parent[v] = u;
            }
        }
    }

    for (int i = 1; i < nodes; i++) 
    {
        cout << parent[i] << " - " << i << endl;
    }
}

int main() 
{
    int nodes = 5;
    vector<vector<pair<int, int>>> adj(nodes);
    adj[0] = {{1, 2}, {3, 6}};
    adj[1] = {{0, 2}, {2, 3}, {3, 8}, {4, 5}};
    adj[2] = {{1, 3}, {4, 7}};
    adj[3] = {{0, 6}, {1, 8}};
    adj[4] = {{1, 5}, {2, 7}};

    prims(nodes, adj);
    return 0;
}

File: Algorithms Advanced/others/GraphAlgorithms/Python/Prim.py
import heapq

def prims(nodes, adj):
    key = [float('inf')] * nodes
    parent = [-1] * nodes
    inMST = [False] * nodes
    key[0] = 0

    pq = [(0, 0)]

    while pq:
        _, u = heapq.heappop(pq)
        inMST[u] = True

        for v, weight in adj[u]:
            if not inMST[v] and weight < key[v]:
                key[v] = weight
                heapq.heappush(pq, (key[v], v))
                parent[v] = u

    for i in range(1, nodes):
        print(f"{parent[i]} - {i}")

# Example
nodes = 5
adj = {
    0: [(1, 2), (3, 6)],
    1: [(0, 2), (2, 3), (3, 8), (4, 5)],
    2: [(1, 3), (4, 7)],
    3: [(0, 6), (1, 8)],
    4: [(1, 5), (2, 7)]
}
prims(nodes, adj)
